liberum_vault/src/vault.rs

```diff
@@ -0,0 +1,481 @@
+use kameo::message::{Context, Message};
+use kameo::Actor;
+use rusqlite::OptionalExtension;
+use tokio::fs::File;
+use tokio::io;
+use tokio::io::AsyncReadExt;
+use tokio::io::AsyncSeekExt;
+use tokio::io::AsyncWriteExt;
+use tokio::io::BufReader;
+use tokio_rusqlite::Connection;
+use tokio_util::bytes::Bytes;
+use tokio_util::io::ReaderStream;
+use tracing::debug;
+
+use crate::fragment::key::Key;
+use crate::fragment::FragmentInfo;
+
+pub struct Vault {
+    db: Connection,
+    vault_dir_path: PathBuf,
+}
+
+type FragmentData = BoxStream<'static, Result<Bytes, io::Error>>;
+
+pub struct StoreFragment(Option<Key>, FragmentData);
+pub struct LoadFragment(Key);
+
+impl Vault {
+    const DEFAULT_VAULT_DATABASE_NAME: &'static str = "vault.db3";
+    const FRAGMENT_DIR_NAME: &'static str = "fragments";
+    const TEMP_DIR_NAME: &'static str = "temp";
+    const MIN_FRAGMENT_SIZE: u64 = 4096;
+
+    pub async fn new(vault_dir_path: &Path) -> Result<Vault> {
+        Self::ensure_dirs(vault_dir_path).await?;
+
+        let db_path = Self::default_db_path(vault_dir_path);
+        let db = Connection::open(db_path).await?;
+
+        Ok(Vault {
+            db,
+            vault_dir_path: vault_dir_path.to_path_buf(),
+        })
+    }
+
+    pub async fn fragment(path: &Path) -> Result<Vec<FragmentData>> {
+        let file_size = tokio::fs::metadata(path).await?.len();
+        let fragment_sizes = Self::fragment_sizes(file_size);
+        let mut current_pos = 0;
+        let mut result: Vec<FragmentData> = Vec::new();
+
+        for fragment_size in fragment_sizes {
+            let mut f = File::open(path).await?;
+            f.seek(std::io::SeekFrom::Start(current_pos)).await?;
+
+            let buf_reader = BufReader::new(f);
+            let reader_stream = ReaderStream::new(buf_reader.take(fragment_size));
+
+            result.push(reader_stream.boxed());
+            current_pos += fragment_size;
+        }
+
+        Ok(result)
+    }
+
+    fn fragment_sizes(target: u64) -> Vec<u64> {
+        let mut fragment_sizes = Vec::new();
+        let mut current_size = cmp::max(Self::MIN_FRAGMENT_SIZE, Self::power_2_upto(target));
+        let mut current_target = target;
+
+        while current_target != 0 {
+            current_size = cmp::max(
+                Self::MIN_FRAGMENT_SIZE,
+                Self::power_2_desc_from_power_2(current_size, current_target),
+            );
+
+            fragment_sizes.push(current_size);
+            current_target = current_target.saturating_sub(current_size);
+        }
+
+        fragment_sizes
+    }
+
+    fn power_2_upto(limit: u64) -> u64 {
+        let powers_from_1 = successors(Some(1u64), |&n| Some(n * 2))
+            .take_while(|x| x <= &limit)
+            .last();
+
+        once(0u64).chain(powers_from_1).last().unwrap()
+    }
+
+    fn power_2_desc_from_power_2(start: u64, limit: u64) -> u64 {
+        assert!(Self::is_power_of_2(start));
```

Comment:
These functions used to fragment into power of 2 sizes seem overly complex and computationally heavy. The power of 2 are calculated for the whole vectors multiple times only to drop all but one of them. 
Although as long as it works, it's fine for now.

Reply:
I think I do not fully understand. `power_2_desc_from_power_2` in most cases just divides by 2 single time and returns the value. So this iterator here will probably run 1-2 times in most cases and then the .last() element will be selected. It is just another way of writing a simple while loop. Not really better way to do this (note that this iterator -- successors -- is iterating from start and descending. So it will be something like 8192 -> ok 8192 is still larger than the limit -> 4096 -- ok 4096 is than the limits so we return it in last.